### PR TITLE
Classification model pipeline

### PR DESCRIPTION
**Pull request by vessemer**
Pipeline to train and predict by the model was provided along with an interface for `classification_model`. 

## Reference to official issue
#131 

## Metrics:
Model' CPM score was described in PR #292:

[CPM](https://doi.org/10.1109/TMI.2010.2072789) over 10-Fold cross validation:

| 0.125 | 0.25 | 0.5 | 1 | 2 | 4 | 8 | Score ([CPM](https://doi.org/10.1109/TMI.2010.2072789)) |
|:-----:| ----- | ----- | ----- | ----- | ----- | ----- |:-----:|
| 0.595 | 0.670 | 0.731 |  0.793 |  0.835 | 0.868 | 0.887 | 0.76 |   

froc_3dlrcnn  

| Item | Config |
| :---: | :----: |
| GPU | NVIDIA TITAN X |
| GPU memory usage | 12GiB for batch size 32 | 

## CLA
- [x] I have signed the CLA; if other committers are in the commit history, they have signed the CLA as well

### CODE REVIEW

**Comment by vessemer** [comment]
Currently, I'm working on the training part adjustment of `grt123` algorithm.

File: prediction/src/algorithms/classify/src/classification_model.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    """
+
+    Args:
+        init_model (bool | str): whether to initialise the model.
+            If str, then the model will be loaded from the init_model
+            path.
+        pull_size (int): maximum amount of batches allowed to be stored in RAM.
+    """
+
+    @abc.abstractmethod
```
Comment (reubano) [inline prediction/src/algorithms/classify/src/classification_model.py]: what does `abstractmethod` do in this case? Python docs say, 

> Using this decorator requires that the class’s metaclass is ABCMeta or is derived from it. 

Also, why not use a new style class, i.e., `class ClassificationModel(object):`

**Comment by reubano** [comment]
Restarted the travis build since the error occurred before even reaching the tests.

**Comment by vessemer** [comment]
Okh, [this](https://github.com/concept-to-clinic/concept-to-clinic/blame/413bc74311579bc3eaa9ba9a147eb8bdc937b093/prediction/src/preprocess/preprocess_ct.py#L167) line along with removal [this](https://github.com/concept-to-clinic/concept-to-clinic/blame/fc9137c2c110f035cdd068bf4b2a5ec479357fc2/prediction/src/preprocess/preprocess_ct.py#L111) one makes me suffer for a while and force to code and tests updates. Unfortunately, at the moment of PR #272 I did not have the opportunity to provide a review, now to be clear:
First: `ndarray` coordinates of the real world point should be computed in this way: `(point - origin) / spacing`and not [`(point - origin) * spacing`](https://github.com/concept-to-clinic/concept-to-clinic/blame/fc9137c2c110f035cdd068bf4b2a5ec479357fc2/prediction/src/preprocess/preprocess_ct.py#L111), where `spacing` is the shape of one voxel in real-world units. 
Second: `meta.spacing` stores information of aforementioned voxel's shape, and if we apply affine transformation such as zoom, then `meta.spacing` should changes too, which is not the previous behaviour. The way that `preprocess_ct.PreprocessCT` handle `spacing` parameter is to zoom an ndarray exactly to this new `spacing` by setting `zoom_fctr` to be `old_spacing / new_spacing` this means that we need no more to store the `old_spacing` in `meta` and at the same time it's enough to store only current spacing for real-world - ndarray coordinates translations. That's why [this](https://github.com/concept-to-clinic/concept-to-clinic/blame/fc9137c2c110f035cdd068bf4b2a5ec479357fc2/prediction/src/preprocess/preprocess_ct.py#L111) line is important.
This mistakes were propagated to the tests in a way of coordinates picking.

Quick check:
tmp

File: prediction/src/algorithms/classify/src/classification_model.py (hunk 1 above)
Comment (vessemer) [reply prediction/src/algorithms/classify/src/classification_model.py]: Indeed, thanks :)

**Comment by lamby** [comment]
@vessemer Can you rework your previous comment into the code itself? Remarks here are really really useful (!) but it won't be clear to anyone following the codebase later, you see...

**Comment by vessemer** [comment]
@lamby, good point, done :)

**Comment by lamby** [comment]
Thanks!

File: prediction/src/tests/test_classification.py [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
-
+# def test_classify_predict_load(metaimage_path, model_path):
+#     assert not trained_model.predict(metaimage_path, [], model_path)
+#
+#
+# def test_classify_dicom(dicom_paths, nodule_locations, model_path):
+#     predicted = trained_model.predict(dicom_paths[0], nodule_locations, model_path)
+#     assert predicted
+#     assert 0 <= predicted[0]['p_concerning'] <= 1
+#
```
Comment (reubano) [inline prediction/src/tests/test_classification.py]: Why comment out these tests? Anything else needed to keep them passing?

Comment (vessemer) [reply prediction/src/tests/test_classification.py]: Not at all, just forget to uncomment them, done [here](https://github.com/concept-to-clinic/concept-to-clinic/pull/304)

**Pull request merged**